x-pack/plugin/enrich/src/main/java/org/elasticsearch/xpack/enrich/EnrichCache.java

```diff
@@ -43,11 +49,11 @@
  */
 public class EnrichCache {
 
-    protected final Cache<CacheKey, CompletableFuture<SearchResponse>> cache;
+    protected final Cache<CacheKey, CompletableFuture<List<Map<?, ?>>>> cache;
```

Comment:
I was thinking, maybe we store the enrich hits as byte[]? And then upon retrieval we parse it into `List<Map<?, ?>>`. 

We already do some work here just before storing a cache entry (recursively make an immutable copy of the search hits (just to catch anyone trying to corrupt the cache) ) and when serving from this cache (recursively make a mutable copy of the search hits). 

I think if we cache the byte[] representation of search hits we save some heap space. At the cost of some additional operations when serving from cache or adding a new entry to cache. But that should be ok, since we already do some extra work making deep copies.

Reply:
It is hard to guess for me what the right tradeoff is in terms of the cache size vs compute cost here. Even a deep copy will be far cheaper than deserialisation that's for sure. I think instinctively I'd rather go for a deep copy because that saves memory after retrieval because all the strings (at least the map keys) will be shared instead of recreated on parsing over and over.

Reply:
I see, this makes sense. Let's stick with creating a deep copy each time an entry from the cache is used.